core/src/main/scala/org/apache/spark/executor/ExecutorMetricsPoller.scala

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.executor
+
+import java.lang.Long.{MAX_VALUE => LONG_MAX_VALUE}
+import java.util.concurrent.{ConcurrentHashMap, TimeUnit}
+import java.util.concurrent.atomic.{AtomicLong, AtomicLongArray}
+
+import scala.collection.mutable.HashMap
+
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.internal.Logging
+import org.apache.spark.memory.MemoryManager
+import org.apache.spark.metrics.ExecutorMetricType
+import org.apache.spark.util.{ThreadUtils, Utils}
+
+/**
+ * :: DeveloperApi ::
+ * A class that polls executor metrics, and tracks their peaks per task and per stage.
+ * Each executor keeps an instance of this class.
+ * The poll method polls the executor metrics, and is either run in its own thread or
+ * called by the executor's heartbeater thread, depending on configuration.
+ * The class keeps two ConcurrentHashMaps that are accessed (via its methods) by the
+ * executor's task runner threads concurrently with the polling thread. One thread may
+ * update one of these maps while another reads it, so the reading thread may not get
+ * the latest metrics, but this is ok.
+ *
+ * @param memoryManager the memory manager used by the executor.
+ * @param pollingInterval the polling interval in milliseconds.
+ */
+@DeveloperApi
+private[spark] class ExecutorMetricsPoller(
```

Comment:
Seems like a lot of threading and concurrency are added here. How much is the overhead? Is there a number?

Reply:
As I explained, there is only ever one thread that polls executor metrics, either a dedicated polling thread, or the heartbeater thread. So, yes, there is possibly one more thread than before. However, I don't think that by itself adds much overhead. There are two ConcurrentHashMaps that we keep here, that the thread that polls updates in a bulk operation while task runner threads may concurrently read from them; this is what CHMs are designed to do well with little overhead. Task runner threads also put an entry in the CHMs when a task starts and updates an entry or removes it when the task ends. This requires synchronization but the operation is short.

Reply:
So I just asked if there was an overhead analysis and if there is a number like 10-20% overhead. If you are saying that you have done some testing on an actual cluster with some large/normal/low workload and concluded that the overhead isn't that much to even being noticeable, I'm totally fine. It is for guiding users to what expect if they want to use this.

Reply:
I have done testing on an actual cluster. I ran a few applications with and without faster polling (100ms interval). In both cases, I used 1 second for the heartbeat interval. There was no noticeable difference to the task times in the executors or to the stage times. However, the workload is small and the tasks took ~15s to ~1 minute.